File: api/src/modules/weather/weather.service.ts
import { Injectable, HttpException } from "@nestjs/common";
import { ConfigService } from "@nestjs/config";
import { HttpService } from "@nestjs/axios";
import { firstValueFrom } from "rxjs";

export interface WeatherCity {
  id: string;
  name: string;
  query: string;
  latitude: number;
  longitude: number;
}

@Injectable()
export class WeatherService {
  private readonly apiKey: string;
  private readonly baseUrl = "https://api.openweathermap.org/data/2.5";
  private readonly defaultCity = "Santa Cruz de la Sierra,BO";

  private readonly cities: WeatherCity[] = [
    { id: "santa-cruz", name: "Santa Cruz de la Sierra", query: "Santa Cruz de la Sierra,BO", latitude: -17.7833, longitude: -63.1821 },
    { id: "la-paz", name: "La Paz", query: "La Paz,BO", latitude: -16.5000, longitude: -68.1500 },
    { id: "cochabamba", name: "Cochabamba", query: "Cochabamba,BO", latitude: -17.3895, longitude: -66.1568 },
    { id: "sucre", name: "Sucre", query: "Sucre,BO", latitude: -19.0333, longitude: -65.2625 },
    { id: "potosi", name: "Potosí", query: "Potosí,BO", latitude: -19.5836, longitude: -65.7531 },
    { id: "oruro", name: "Oruro", query: "Oruro,BO", latitude: -17.9833, longitude: -67.1500 },
    { id: "tarija", name: "Tarija", query: "Tarija,BO", latitude: -21.5355, longitude: -64.7296 },
    { id: "trinidad", name: "Trinidad", query: "Trinidad,BO", latitude: -14.8333, longitude: -64.9000 },
    { id: "cobija", name: "Cobija", query: "Cobija,BO", latitude: -11.0267, longitude: -68.7692 },
    { id: "el-alto", name: "El Alto", query: "El Alto,BO", latitude: -16.5044, longitude: -68.1631 },
  ];

  constructor(
    private configService: ConfigService,
    private httpService: HttpService,
  ) {
    this.apiKey = this.configService.get<string>("OPENWEATHER_API_KEY") || "";
  }

  getCities(): WeatherCity[] {
    return this.cities;
  }

  private assertApiKeyConfigured() {
    if (!this.apiKey) {
      throw new HttpException("Weather service unavailable: API key not configured", 503);
    }
  }

  private resolveQuery(cityId?: string): string {
    const city = this.cities.find((c) => c.id === cityId);
    return city ? city.query : this.defaultCity;
  }

  async getCurrent(cityId?: string) {
    this.assertApiKeyConfigured();
    try {
      const { data } = await firstValueFrom(
        this.httpService.get(`${this.baseUrl}/weather`, {
          params: {
            q: this.resolveQuery(cityId),
            appid: this.apiKey,
            units: "metric",
            lang: "es",
          },
        }),
      );

      return {
        temperature: Math.round(data.main.temp),
        feelsLike: Math.round(data.main.feels_like),
        humidity: data.main.humidity,
        description: data.weather[0].description,
        icon: data.weather[0].icon,
        windSpeed: data.wind.speed,
        city: data.name,
      };
    } catch (error) {
      throw new HttpException("Error fetching weather data", 502);
    }
  }

  async getForecast(cityId?: string) {
    this.assertApiKeyConfigured();
    try {
      const { data } = await firstValueFrom(
        this.httpService.get(`${this.baseUrl}/forecast`, {
          params: {
            q: this.resolveQuery(cityId),
            appid: this.apiKey,
            units: "metric",
            lang: "es",
            cnt: 40,
          },
        }),
      );

      const daily = data.list
        .filter((_: any, i: number) => i % 8 === 0)
        .map((item: any) => ({
          date: item.dt_txt.split(" ")[0],
          temperature: Math.round(item.main.temp),
          tempMin: Math.round(item.main.temp_min),
          tempMax: Math.round(item.main.temp_max),
          description: item.weather[0].description,
          icon: item.weather[0].icon,
          humidity: item.main.humidity,
        }));

      return { city: data.city.name, forecast: daily };
    } catch (error) {
      throw new HttpException("Error fetching forecast data", 502);
    }
  }
}

File: app/lib/features/weather/data/weather_service.dart
import 'package:dio/dio.dart';
import '../../../../config/api_constants.dart';

class WeatherCity {
  final String id;
  final String name;
  final double latitude;
  final double longitude;

  const WeatherCity({
    required this.id,
    required this.name,
    required this.latitude,
    required this.longitude,
  });

  factory WeatherCity.fromJson(Map<String, dynamic> json) {
    return WeatherCity(
      id: json['id']?.toString() ?? '',
      name: json['name'] ?? '',
      latitude: double.tryParse(json['latitude']?.toString() ?? '') ?? 0,
      longitude: double.tryParse(json['longitude']?.toString() ?? '') ?? 0,
    );
  }
}

class WeatherCurrent {
  final int temperature;
  final int feelsLike;
  final int humidity;
  final String description;
  final String? icon;
  final double windSpeed;
  final String city;

  const WeatherCurrent({
    required this.temperature,
    required this.feelsLike,
    required this.humidity,
    required this.description,
    this.icon,
    required this.windSpeed,
    required this.city,
  });

  factory WeatherCurrent.fromJson(Map<String, dynamic> json) {
    return WeatherCurrent(
      temperature: json['temperature'] is num
          ? (json['temperature'] as num).round()
          : int.tryParse(json['temperature']?.toString() ?? '') ?? 0,
      feelsLike: json['feelsLike'] is num
          ? (json['feelsLike'] as num).round()
          : int.tryParse(json['feelsLike']?.toString() ?? '') ?? 0,
      humidity: json['humidity'] is num
          ? (json['humidity'] as num).round()
          : int.tryParse(json['humidity']?.toString() ?? '') ?? 0,
      description: json['description'] ?? '',
      icon: json['icon'],
      windSpeed: double.tryParse(json['windSpeed']?.toString() ?? '') ?? 0,
      city: json['city'] ?? '',
    );
  }
}

class WeatherForecastDay {
  final DateTime date;
  final int temperature;
  final int tempMin;
  final int tempMax;
  final String description;
  final String? icon;
  final int humidity;

  const WeatherForecastDay({
    required this.date,
    required this.temperature,
    required this.tempMin,
    required this.tempMax,
    required this.description,
    this.icon,
    required this.humidity,
  });

  factory WeatherForecastDay.fromJson(Map<String, dynamic> json) {
    return WeatherForecastDay(
      date: DateTime.tryParse(json['date'] ?? '') ?? DateTime.now(),
      temperature: json['temperature'] is num
          ? (json['temperature'] as num).round()
          : int.tryParse(json['temperature']?.toString() ?? '') ?? 0,
      tempMin: json['tempMin'] is num
          ? (json['tempMin'] as num).round()
          : int.tryParse(json['tempMin']?.toString() ?? '') ?? 0,
      tempMax: json['tempMax'] is num
          ? (json['tempMax'] as num).round()
          : int.tryParse(json['tempMax']?.toString() ?? '') ?? 0,
      description: json['description'] ?? '',
      icon: json['icon'],
      humidity: json['humidity'] is num
          ? (json['humidity'] as num).round()
          : int.tryParse(json['humidity']?.toString() ?? '') ?? 0,
    );
  }
}

class WeatherForecast {
  final String city;
  final List<WeatherForecastDay> days;

  const WeatherForecast({required this.city, required this.days});

  factory WeatherForecast.fromJson(Map<String, dynamic> json) {
    final list = (json['forecast'] as List?)?.whereType<Map<String, dynamic>>() ?? [];
    return WeatherForecast(
      city: json['city'] ?? '',
      days: list.map(WeatherForecastDay.fromJson).toList(),
    );
  }
}

class WeatherService {
  final Dio _dio;

  WeatherService(this._dio);

  Future<List<WeatherCity>> getCities() async {
    final response = await _dio.get(ApiConstants.weatherCities);
    final data = response.data['data'] ?? response.data;
    return (data as List)
        .whereType<Map<String, dynamic>>()
        .map(WeatherCity.fromJson)
        .toList();
  }

  Future<WeatherCurrent> getCurrent({String? cityId}) async {
    final response = await _dio.get(
      ApiConstants.weatherCurrent,
      queryParameters: cityId != null && cityId.isNotEmpty ? {'city': cityId} : null,
    );
    final data = response.data['data'] ?? response.data;
    final current = data is Map<String, dynamic> ? data : (data['current'] ?? const {});
    return WeatherCurrent.fromJson(current);
  }

  Future<WeatherForecast> getForecast({String? cityId}) async {
    final response = await _dio.get(
      ApiConstants.weatherForecast,
      queryParameters: cityId != null && cityId.isNotEmpty ? {'city': cityId} : null,
    );
    final data = response.data['data'] ?? response.data;
    return WeatherForecast.fromJson(data is Map<String, dynamic> ? data : const {});
  }
}